File: src/clj/io/github/amiorin/walter/options.clj
(ns io.github.amiorin.walter.options
  (:require
   [big-config.render :as render]
   [big-config.workflow :as workflow]))

(def ^:private oci {::workflow/params {:provider-compute "oci"
                                       :oci-config-file-profile "DEFAULT"
                                       :oci-subnet-id "ocid1.subnet.oc1.eu-frankfurt-1.aaaaaaaaotya32pihejgi25vrdfnjda3qg52kpsjnd7od5oiqifbsi4rqqma"
                                       :oci-compartment-id "ocid1.tenancy.oc1..aaaaaaaal4wmmpzv2fzkdz2vrfdizywgzjid6dqlgcankrrr7jyydo7ozb3a"
                                       :oci-availability-domain "xTQn:EU-FRANKFURT-1-AD-1"
                                       :oci-display-name "my-ampere-instance"
                                       :oci-shape "VM.Standard.A1.Flex"
                                       :oci-ocpus 1
                                       :oci-memory-in-gbs 6
                                       :oci-boot-volume-size-in-gbs 50
                                       :oci-boot-volume-vpus-per-gb 30
                                       :oci-ssh-authorized-keys "~/.ssh/id_ed25519.pub"}})

(def ^:private s3 {::workflow/params {:provider-backend "s3"
                                      :s3-bucket "tf-state-251213589273-eu-west-1"
                                      :s3-region "eu-west-1"}})

(def walter (merge-with merge oci s3
                        {::render/profile "walter"
                         ::workflow/params {:package "walter"}}))

(def bb walter)

(comment
  (-> bb))

File: src/walter/options.py
"""Walter profiles and the active profile."""
from __future__ import annotations

from typing import Any

from big_config.core import Opts

from .interop import PARAMS, PROFILE, sync_aliases

Profile = dict[str, Any]


def compose(*layers: Profile) -> Opts:
    """Compose layers like Clojure's ``merge-with merge``: params merge shallowly."""
    params: dict[str, Any] = {}
    profile: str | None = None
    for layer in layers:
        if "profile" in layer:
            profile = layer["profile"]
        params = {**params, **(layer.get("params") or {})}
    return sync_aliases({PROFILE: profile, PARAMS: params})


base: Profile = {
    "profile": "walter",
    "params": {
        "package": "walter",
        "compute-pubkey": "REPLACE_ME",
        "provider-compute": "hcloud",
        "provider-backend": "r2",
        "compute-prevent-destroy": True,
        "s3-bucket": "REPLACE_ME",
        "s3-region": "REPLACE_ME",
        "r2-bucket": "REPLACE_ME",
        "r2-endpoint": "REPLACE_ME",
        "r2-access-key-id": "REPLACE_ME",
        "r2-secret-access-key": "REPLACE_ME",
        "oci-config-file-profile": "REPLACE_ME",
        "oci-subnet-id": "REPLACE_ME",
        "oci-compartment-id": "REPLACE_ME",
        "oci-availability-domain": "REPLACE_ME",
        "oci-display-name": "walter",
        "oci-shape": "VM.Standard.A1.Flex",
        "oci-ocpus": 1,
        "oci-memory-in-gbs": 4,
        "oci-boot-volume-size-in-gbs": 50,
        "oci-boot-volume-vpus-per-gb": 30,
        "oci-ssh-authorized-keys": "REPLACE_ME",
        "hcloud-name": "walter",
        "hcloud-image": "ubuntu-24.04",
        "hcloud-server-type": "cx23",
        "hcloud-location": "hel1",
        "hcloud-ssh-keys": "REPLACE_ME",
        "hcloud-token": "REPLACE_ME",
        "digitalocean-name": "walter",
        "digitalocean-region": "ams3",
        "digitalocean-size": "s-1vcpu-1gb-35gb-intel",
        "digitalocean-image": "ubuntu-25-10-x64",
        "digitalocean-vpc-uuid": "REPLACE_ME",
        "digitalocean-ssh-keys": "REPLACE_ME",
        "do-token": "REPLACE_ME",
        "no-infra-compute-ip": "REPLACE_ME",
        "no-infra-compute-user": "ubuntu",
        "no-infra-compute-sudoer": "root",
        "no-infra-compute-uid": "1000",
    },
}

walter: Opts = compose(base)
bb: Opts = walter
